keps/sig-cluster-lifecycle/wgs/20191001-component-config-validation-interface.md

```diff
@@ -0,0 +1,177 @@
+
+
+## Motivation
+
+With the advent of component configs, it became possible for cluster lifecycle tools (such as kubeadm) to generate and supply provisioned components with config files in YAML or JSON format. This is gradually increasing in popularity as maintaining a well structured and serialized in a text file config is far easier than maintaining a large set of command line arguments.
+
+This, however, poses some problems. For example:
+- The supplied config may be in an unsupported (newer or older) or unrecognized format.
+- The config file may be corrupted due to disk, network, program or user error.
+- A required system setting or module may not be setup properly or available, hence, prohibiting component operation with the current config.
+
+Unfortunately, due to the nature of Kubernetes, such kinds of failures might not be detected by cluster lifecycle tools early enough and users might not be warned in time, that their configs are invalid.  
+Currently, some cluster lifecycle tools vendor in and use Go code from the target components themselves to be able to verify component configs early enough during the provisioning operation, so that meaningful error messages are displayed in time directly to the user. This, however, poses its own problems as such pieces of code can deviate from the actual shipped component and can lead to false positives and false negatives.  
+Hence, it’s best that validation is performed by the component itself in a one off operation.
+
+
+### Goals
+
+The main goal is to provide a stable standard command line interface for components that can be used to validate component configs and retrieve appropriate errors and warnings.
+
+
+### Non-Goals
+
+To impose the implementation of this proposal on all Kubernetes components, that provide component config. Implementing this interface is entirely optional.
+
+
+## Proposal
+
+The proposal is to introduce a new standardized sub-command for each component that implements the interface. This command can then be called by invoking the component’s executable or creating a container with it.
+
+
+### The “experimental-config-validate“ sub-command
+
+The new sub-command will be known as “experimental-config-validate”. Upon graduation to beta and GA, this sub-command will be renamed to just “config-validate”.
+
+It must support a couple of mandatory flags:
+- “--help” or “-h” - if specified, will dump usage information of the sub-command and exit without doing any config validation.
+- “--file” or “-f” - supplies a YAML or JSON file containing a component config to be validated.
+
+In all implementations of “experimental-config-validate”, the “--file” flag can be specified multiple times. If a component does not support multi-file config, supplying multiple “--file” values must return an appropriate error to the user without performing any validation.  
+If no “--file” or “--help” switches are supplied, the command reads the config to be validated from the standard input.
```

Comment:
I'm assuming we plan on allowing a stream of documents on STDIN?

Reply:
This does add another input vector to most components, which don't do anything (AFAIK) with STDIN today. We aren't monopolizing STDIN (controlled via a flag) but worth asking what the impact of that is.

Reply:
Piping stuff is handy. That's why I included it. After all this would be handy too:
```
kubectl get cm kube-proxy -n kube-system -o template --template '{{index .data "config.conf" }}' | kube-proxy experimental-config-validate
```